distsql: improve hashjoin performance

jordanlewis opened the pull request:
- add a benchmark for the spilling hashjoiner
- reset the disk iterator rather than creating a new one every probeRow
- fix small flushing bug that caused a flush on an empty batch

```
name                                old time/op    new time/op    delta
HashJoiner/spill=true/rows=256-8      11.7ms ± 3%     2.2ms ± 3%   -81.43%  (p=0.000 n=10+8)
HashJoiner/spill=true/rows=4096-8     89.9ms ± 2%    18.4ms ± 1%   -79.58%  (p=0.000 n=10+9)
HashJoiner/spill=true/rows=65536-8     1.02s ± 2%     0.28s ± 2%   -72.16%  (p=0.000 n=9+10)

name                                old speed      new speed      delta
HashJoiner/spill=true/rows=256-8     175kB/s ± 3%   941kB/s ± 3%  +437.86%  (p=0.000 n=10+8)
HashJoiner/spill=true/rows=4096-8    364kB/s ± 2%  1780kB/s ± 2%  +389.01%  (p=0.000 n=10+10)
HashJoiner/spill=true/rows=65536-8   517kB/s ± 1%  1850kB/s ± 2%  +258.06%  (p=0.000 n=9+10)

name                                old alloc/op   new alloc/op   delta
HashJoiner/spill=true/rows=256-8       500kB ± 0%      46kB ± 0%   -90.82%  (p=0.000 n=9+9)
HashJoiner/spill=true/rows=4096-8     8.04MB ± 0%    0.79MB ± 0%   -90.16%  (p=0.000 n=9+9)
HashJoiner/spill=true/rows=65536-8     129MB ± 0%      13MB ± 0%   -90.09%  (p=0.000 n=10+9)

name                                old allocs/op  new allocs/op  delta
HashJoiner/spill=true/rows=256-8       2.88k ± 0%     0.83k ± 0%   -71.10%  (p=0.000 n=10+10)
HashJoiner/spill=true/rows=4096-8      45.4k ± 0%     12.7k ± 0%   -72.13%  (p=0.000 n=10+10)
HashJoiner/spill=true/rows=65536-8      726k ± 0%      202k ± 0%   -72.19%  (p=0.000 n=10+10)
```

jordanlewis committed:
distsqlrun: hashjoiner benchmark can spill to disk

Release note: None

jordanlewis committed:
distsqlrun: improve performance of hash join

Previously, the hash joiner created a new hash row container iterator on
every invocation of probeRow. This was wasteful. Now, we reuse the
iterator and reset it on each invocation.

Also, a  small bug was causing the disk map to flush even if it had no
rows to flush. Correct this.

```
name                                old time/op    new time/op    delta
HashJoiner/spill=true/rows=256-8      11.7ms ± 3%     2.2ms ± 3%   -81.43%  (p=0.000 n=10+8)
HashJoiner/spill=true/rows=4096-8     89.9ms ± 2%    18.4ms ± 1%   -79.58%  (p=0.000 n=10+9)
HashJoiner/spill=true/rows=65536-8     1.02s ± 2%     0.28s ± 2%   -72.16%  (p=0.000 n=9+10)

name                                old speed      new speed      delta
HashJoiner/spill=true/rows=256-8     175kB/s ± 3%   941kB/s ± 3%  +437.86%  (p=0.000 n=10+8)
HashJoiner/spill=true/rows=4096-8    364kB/s ± 2%  1780kB/s ± 2%  +389.01%  (p=0.000 n=10+10)
HashJoiner/spill=true/rows=65536-8   517kB/s ± 1%  1850kB/s ± 2%  +258.06%  (p=0.000 n=9+10)

name                                old alloc/op   new alloc/op   delta
HashJoiner/spill=true/rows=256-8       500kB ± 0%      46kB ± 0%   -90.82%  (p=0.000 n=9+9)
HashJoiner/spill=true/rows=4096-8     8.04MB ± 0%    0.79MB ± 0%   -90.16%  (p=0.000 n=9+9)
HashJoiner/spill=true/rows=65536-8     129MB ± 0%      13MB ± 0%   -90.09%  (p=0.000 n=10+9)

name                                old allocs/op  new allocs/op  delta
HashJoiner/spill=true/rows=256-8       2.88k ± 0%     0.83k ± 0%   -71.10%  (p=0.000 n=10+10)
HashJoiner/spill=true/rows=4096-8      45.4k ± 0%     12.7k ± 0%   -72.13%  (p=0.000 n=10+10)
HashJoiner/spill=true/rows=65536-8      726k ± 0%      202k ± 0%   -72.19%  (p=0.000 n=10+10)
```

Release note (performance improvement): improve performance of hash join

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24577)
<!-- Reviewable:end -->

petermattis commented:
Nice speedup!

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-:-L9VAtZV02M7QBMo4rzp:buerb8d)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
PTAL.

The original patchset had a bug. You still have to recreate the underlying iterator if you've written any marks to disk. This deficiency can be solved by a cache of written marks, which we should have memory space for because once we spill to disk we abandon our memory hash table. This work is tracked in #24580.

petermattis commented:
LGTM, though I only skimmed it. @asubiotto or @arjunravinarayan will have to give this a thorough review. 

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 531 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9aMKtT-Tdo1-UC_Wsx:-L9aMKtT-Tdo1-UC_Wsy:b56tcxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/13002f264a7e372d24a67f57159c95d6143eae90/pkg/sql/distsqlrun/hash_row_container.go#L531)):*
> ```Go
> 	if !i.bufferedRows.Empty() {
> 		// We have to recreate our iterator if we need to flush marks to disk.
> 		// TODO(jordan): do this less by keeping a cache of written marks.
> ```

Is recreating the iterator showing up as expensive? RocksDB also has the concept of a "tailing" iterator which sees the latest database state instead of a snapshot.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-:-L9aNHN48Puym-s0qSul:b-rweoyg)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 531 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9aMKtT-Tdo1-UC_Wsx:-L9b5D4f9EAFSTdtUJx1:b753kei) ([raw file](https://github.com/cockroachdb/cockroach/blob/13002f264a7e372d24a67f57159c95d6143eae90/pkg/sql/distsqlrun/hash_row_container.go#L531)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is recreating the iterator showing up as expensive? RocksDB also has the concept of a "tailing" iterator which sees the latest database state instead of a snapshot.
</blockquote></details>

Yes, the reason the benchmark improved so much is that it never has to recreate/delete the iterator anymore. 

A similar benchmark of a join that needs to use marking, like EXCEPT ALL etc, would not see as big of an improvement until we complete that TODO.

I'll have to look into the tailing iterator. That might be exactly what we need.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsqlrun/disk_row_container.go, line 228 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f45LG4hHfQ4YtYTSn:-L9f45LG4hHfQ4YtYTSo:b52lb4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/disk_row_container.go#L228)):*
> ```Go
> 
> func (d diskRowIterator) Close() {
> 	if d.SortedDiskMapIterator != nil {
> ```

Is this ever not true?

---

*[pkg/sql/distsqlrun/disk_row_container.go, line 227 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f9Ri6BLfVQkMb4Z4i:-L9f9Ri6BLfVQkMb4Z4j:bwwlc96) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9b7aeaed30cc6a661f0a70dd9a1d502d0fe848d/pkg/sql/distsqlrun/disk_row_container.go#L227)):*
> ```Go
> }
> 
> func (r diskRowIterator) Close() {
> ```

I wonder why I chose `r` in `Row`, feel free to use whatever you prefer (maybe `i`)

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 34 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f2KlM5PBLUDLihpG_:-L9f2KlM5PBLUDLihpGa:b-c80w0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L34)):*
> ```Go
> type rowMarkerIterator interface {
> 	rowIterator
> 	Reset(ctx context.Context, row sqlbase.EncDatumRow) error
> ```

This probably deserves a comment

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 263 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f17Pq9gK5PxOlbKCN:-L9f17Pq9gK5PxOlbKCO:bf5cqla) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L263)):*
> ```Go
> 	ctx context.Context, row sqlbase.EncDatumRow, probeEqCols columns,
> ) (rowMarkerIterator, error) {
> ```

nit: extra line

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 480 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f4gN-CBmjy7yHdWcq:-L9f4gN-CBmjy7yHdWcr:b7av1wa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L480)):*
> ```Go
> 	ctx context.Context, row sqlbase.EncDatumRow, probeEqCols columns,
> ) (rowMarkerIterator, error) {
> 	ret := hashDiskRowBucketIterator{
> ```

nit: you took the address of the iterator here in `hashMemRowContainer.NewBucketIterator`

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 529 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f61f_0FPaMXCsHPB2:-L9f61f_0FPaMXCsHPB3:bjznanh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L529)):*
> ```Go
> 	}
> 	i.encodedEqCols = append(i.encodedEqCols[:0], encoded...)
> 	if !i.bufferedRows.Empty() {
> ```

Your comment on `Empty` in `disk_map.go` made me stop and think here. The batch could be empty right after a flush and we wouldn't recreate an iterator although we would need to. Is that correct? Also posted a comment in `disk_map.go`

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 580 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f3t8U7ss8uAFG6v7i:-L9f3t8U7ss8uAFG6v7j:b8pkn5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L580)):*
> ```Go
> // Close implements the rowIterator interface.
> func (i *hashDiskRowBucketIterator) Close() {
> 	i.diskRowIterator.Close()
> ```

I'm confused. Do we need this `Close` since `diskRowIterator` is embedded in `hashDiskRowBucketIterator`?

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 100 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f-smP7nzXMDi5A6pi:-L9f-smP7nzXMDi5A6pj:b1ldtiu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hashjoiner.go#L100)):*
> ```Go
> 	forcedStoredSide *joinSide
> 
> 	bucketIterator rowMarkerIterator
> ```

I would add a comment here

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 205 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f0YfT98jogqhMqa42:-L9f0YfT98jogqhMqa43:b-1q5kzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hashjoiner.go#L205)):*
> ```Go
> 	defer h.rows[rightSide].Close(ctx)
> 
> 	defer func() {
> ```

Maybe move this down to before the call to `probePhase`.

---

*[pkg/sql/distsqlrun/hashjoiner_test.go, line 1439 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9ewu5s0QUrBEW1GeSk:-L9ewu5s0QUrBEW1GeSl:bn5jezf) ([raw file](https://github.com/cockroachdb/cockroach/blob/91fa8fcb3305ebf1a5e5ee4efe79bdf383caa947/pkg/sql/distsqlrun/hashjoiner_test.go#L1439)):*
> ```Go
> 				}
> 				b.Run(fmt.Sprintf("rows=%d", numRows), func(b *testing.B) {
> 					tempEngine, err := engine.NewTempEngine(base.DefaultTestTempStorageConfig(st))
> ```

I believe you can reuse this `tempEngine` to avoid recreating it on each `b.Run` since we prefix keys with a unique ID per instance (a new "instance" is created on each `hashjoiner` run).

---

*[pkg/storage/engine/disk_map.go, line 82 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f8E5i0oqyFTBOlXcD:-L9f8E5i0oqyFTBOlXcE:b-8gfo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/storage/engine/disk_map.go#L82)):*
> ```Go
> 	// after a call to Flush().
> 	Flush() error
> 	// Empty returns true if there have been no writes to this batch writer.
> ```

This is not technically true although it's the behavior we want. After a flush we would re-create a batch and `b.batch.Empty()` would return true. We might need to add some flag.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
I'd like the comment additions that @asubiotto requested, but otherwise LGTM.

---

Reviewed 4 of 4 files at r2, 1 of 2 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/hashjoiner_test.go, line 1439 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9ewu5s0QUrBEW1GeSk:-L9fJTUD1dvuQKZP1Ueu:bwquki6) ([raw file](https://github.com/cockroachdb/cockroach/blob/91fa8fcb3305ebf1a5e5ee4efe79bdf383caa947/pkg/sql/distsqlrun/hashjoiner_test.go#L1439)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I believe you can reuse this `tempEngine` to avoid recreating it on each `b.Run` since we prefix keys with a unique ID per instance (a new "instance" is created on each `hashjoiner` run).
</blockquote></details>

I'm worried about performance differences. Would it still be apples-to-apples?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-:-L9fNtn73UNKnoMZRSBE:b-qmp4qb)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/hashjoiner_test.go, line 1439 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9ewu5s0QUrBEW1GeSk:-L9fON3fAzkxcyK6cheo:bdyso3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/91fa8fcb3305ebf1a5e5ee4efe79bdf383caa947/pkg/sql/distsqlrun/hashjoiner_test.go#L1439)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I'm worried about performance differences. Would it still be apples-to-apples?
</blockquote></details>

I was wondering about that. We clear all the keys written on the instance (not engine) `Close` (when the `hashJoiner` finishes its run) so I think it should be fine.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Thanks for the thorough reviews!

---

Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/disk_row_container.go, line 228 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f45LG4hHfQ4YtYTSn:-L9fBhjVAnDVK0DUx_0B:b-3aifb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/disk_row_container.go#L228)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is this ever not true?
</blockquote></details>

Yes, on initialization. The constructor calls `Reset` to avoid some duplicated code.

---

*[pkg/sql/distsqlrun/disk_row_container.go, line 227 at r4](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f9Ri6BLfVQkMb4Z4i:-L9fBehx-0CU3xYK529n:bnika3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9b7aeaed30cc6a661f0a70dd9a1d502d0fe848d/pkg/sql/distsqlrun/disk_row_container.go#L227)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I wonder why I chose `r` in `Row`, feel free to use whatever you prefer (maybe `i`)
</blockquote></details>

We can fix that later. They have to be consistent or else the linter fails.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 34 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f2KlM5PBLUDLihpG_:-L9fD3bO5fqq_X9ZKZFg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L34)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This probably deserves a comment
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 263 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f17Pq9gK5PxOlbKCN:-L9fCN_678iuR_U_J8am:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L263)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: extra line
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 480 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f4gN-CBmjy7yHdWcq:-L9fCQdD65Mjg_EWB9i3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L480)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: you took the address of the iterator here in `hashMemRowContainer.NewBucketIterator`
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 529 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f61f_0FPaMXCsHPB2:-L9fBu4WC3FMrA_PwdKc:b1cmnza) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L529)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Your comment on `Empty` in `disk_map.go` made me stop and think here. The batch could be empty right after a flush and we wouldn't recreate an iterator although we would need to. Is that correct? Also posted a comment in `disk_map.go`
</blockquote></details>

Oh, great catch. I replaced this Empty method with a simple boolean flag on the hashDiskRowBucketIterator itself, which tracks whether we've written any marks since the last time we've created our underlying iterator.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 580 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f3t8U7ss8uAFG6v7i:-L9fC7vV4n-e-rq6lc6M:b87zmgi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hash_row_container.go#L580)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm confused. Do we need this `Close` since `diskRowIterator` is embedded in `hashDiskRowBucketIterator`?
</blockquote></details>

No, I think this was leftover from something else I was trying. Deleted.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 100 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f-smP7nzXMDi5A6pi:-L9fCHRX66AKYTs44uqH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hashjoiner.go#L100)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would add a comment here
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 205 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f0YfT98jogqhMqa42:-L9fE8gz7r-lvIBytAma:b-5q277z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/sql/distsqlrun/hashjoiner.go#L205)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Maybe move this down to before the call to `probePhase`.
</blockquote></details>

Moved it inside of `probePhase`.

---

*[pkg/sql/distsqlrun/hashjoiner_test.go, line 1439 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9ewu5s0QUrBEW1GeSk:-L9flIh491tfOgcVSWYN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91fa8fcb3305ebf1a5e5ee4efe79bdf383caa947/pkg/sql/distsqlrun/hashjoiner_test.go#L1439)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I was wondering about that. We clear all the keys written on the instance (not engine) `Close` (when the `hashJoiner` finishes its run) so I think it should be fine.
</blockquote></details>

Done.

---

*[pkg/storage/engine/disk_map.go, line 82 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9f8E5i0oqyFTBOlXcD:-L9fGApc-7om2UawfdGy:b-6ya215) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0286f67f255531662e670e8429a4abffb14fa32/pkg/storage/engine/disk_map.go#L82)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This is not technically true although it's the behavior we want. After a flush we would re-create a batch and `b.batch.Empty()` would return true. We might need to add some flag.
</blockquote></details>

Deleted this method.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-:-L9flb3_6O1ES_3sQ_JM:b-u9bhow)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just some minor nits that you can ignore

---

Reviewed 1 of 5 files at r5, 3 of 4 files at r7.
Review status: 4 of 5 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 100 at r7](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9k2_sQ6FmnmVNaHQnH:-L9k2_sQ6FmnmVNaHQnI:bquem1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce90b1846d78350c79cdc3e7e883533fa8ae030f/pkg/sql/distsqlrun/hashjoiner.go#L100)):*
> ```Go
> 	forcedStoredSide *joinSide
> 
> 	bucketIterator rowMarkerIterator
> ```

Just double-checking that you added the comment

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 585 at r7](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9k2LhZ9_u-pm1qy8qi:-L9k2LhZ9_u-pm1qy8qj:bf5cqla) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce90b1846d78350c79cdc3e7e883533fa8ae030f/pkg/sql/distsqlrun/hashjoiner.go#L585)):*
> ```Go
> 	ctx context.Context, storedRows hashRowContainer,
> ) (earlyExit bool, _ error) {
> ```

nit: extra line

---

*[pkg/sql/distsqlrun/hashjoiner_test.go, line 1433 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9jzMjhA5mxqxS_Egmx:-L9jzMjhA5mxqxS_Egmy:b-fxen3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5d64b7ab7a05679d6dc9fa40bb75163f31459ce/pkg/sql/distsqlrun/hashjoiner_test.go#L1433)):*
> ```Go
> 	const numCols = 1
> 	for _, spill := range []bool{true, false} {
> 		flowCtx.testingKnobs = TestingKnobs{}
> ```

nit: I think you can remove this line

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-:-L9k1zCl5enVJGAvXTvA:b-6kr594)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 100 at r7](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9k2_sQ6FmnmVNaHQnH:-L9k6YG_0EiXzYeMXVjM:b8jjy1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce90b1846d78350c79cdc3e7e883533fa8ae030f/pkg/sql/distsqlrun/hashjoiner.go#L100)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Just double-checking that you added the comment
</blockquote></details>

Whoops, Done.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 585 at r7](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9k2LhZ9_u-pm1qy8qi:-L9kA7Qu556uRcZ9j8Hj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce90b1846d78350c79cdc3e7e883533fa8ae030f/pkg/sql/distsqlrun/hashjoiner.go#L585)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: extra line
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hashjoiner_test.go, line 1433 at r5](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577#-L9jzMjhA5mxqxS_Egmx:-L9kAAn00wXi3GU3VPBF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5d64b7ab7a05679d6dc9fa40bb75163f31459ce/pkg/sql/distsqlrun/hashjoiner_test.go#L1433)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I think you can remove this line
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24577)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
bors r+

jordanlewis commented:
bors r+

jordanlewis commented:
Flaky tests getting me down!

jordanlewis commented:
bors r+

couchand commented:
preempted by a green button merge
bors r+

jordanlewis commented:
Uploaded a new patch that's rebased on top of the flaky test skips.

jordanlewis commented:
bors r+

arjunravinarayan commented:
And now his watch is ended.
